paddle/fluid/distributed/auto_parallel/spmd_rules/embedding_spmd_rule.cc

```diff
@@ -0,0 +1,157 @@
+/* Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/distributed/auto_parallel/spmd_rules/embedding_spmd_rule.h"
+
+namespace paddle {
+namespace distributed {
+namespace auto_parallel {
+
+using phi::distributed::auto_parallel::str_join;
+
+// step0: verify input args based on embedding logic
+std::pair<std::vector<TensorDistAttr>, std::vector<TensorDistAttr>>
+EmbeddingSPMDRule::InferForward(const std::vector<DistTensorSpec>& input_specs,
+                                const paddle::framework::AttributeMap& attrs) {
+  auto input_specs_size = input_specs.size();
+  PADDLE_ENFORCE_EQ(
+      input_specs_size,
+      2,
+      phi::errors::InvalidArgument(
+          "The size of InputSpec of embedding should be 2, but got [%d].",
+          input_specs_size));
+  auto x_shape = input_specs[0].shape();
+  auto weight_shape = input_specs[1].shape();
+  int x_ndim = x_shape.size();
+  int weight_ndim = weight_shape.size();
+  auto x_dist_attr_src = input_specs[0].dist_attr();
+  auto weight_dist_attr_src = input_specs[1].dist_attr();
+  std::vector<int64_t> x_dims_mapping = x_dist_attr_src.dims_mapping();
+  std::vector<int64_t> weight_dims_mapping =
+      weight_dist_attr_src.dims_mapping();
+
+  PADDLE_ENFORCE_EQ(
+      x_ndim,
+      x_dims_mapping.size(),
+      phi::errors::InvalidArgument(
+          "Mismatch of X's tensor size: [%d] and X's dims_mapping size [%d].",
+          x_ndim,
+          x_dims_mapping.size()));
+  PADDLE_ENFORCE_EQ(
+      weight_ndim,
+      weight_dims_mapping.size(),
+      phi::errors::InvalidArgument(
+          "Mismatch of Y's tensor size: [%d] and Y's dims_mapping size [%d].",
+          weight_ndim,
+          weight_dims_mapping.size()));
+  PADDLE_ENFORCE_EQ(
+      weight_ndim,
+      2,
+      phi::errors::InvalidArgument("Embedding table should have TWO dimension, "
+                                   "but got a tensor with [%d] dimension.",
+                                   weight_ndim));
+
+  int64_t padding_idx = ExtractAttr<int64_t>("padding_idx", attrs);
+  bool sparse = ExtractAttr<bool>("sparse", attrs);
+
+  // determine parallel mode
+  int64_t weight_row_axis_mapping = weight_dims_mapping[0];
+
+  // padding_idx s not supported by c_embedding kernl.
+  // (TODO) might be could reshard as replicated when padding_idx != -1
```

Comment:
hard to understand

Reply:
the implement of  c_embedding kernel  require the padding item is in the of table, which means "padding_idx = -1" and not allow being set by user.

to support vocab parallel of embedding op efficiently, we need to use c_embedding kernel, therefore, we make a precondition check that "padding_idx = -1" when embedding table is sharded by vocab axis.